app/src/main/java/edu/wisc/ece/pinpoint/data/Pin.java

```diff
@@ -0,0 +1,77 @@
+package edu.wisc.ece.pinpoint.data;
+
+import android.util.Log;
+
+import androidx.annotation.NonNull;
+
+import com.google.android.gms.tasks.OnFailureListener;
+import com.google.android.gms.tasks.OnSuccessListener;
+import com.google.firebase.firestore.DocumentReference;
+import com.google.firebase.firestore.FirebaseFirestore;
+
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+
+public class Pin {
+    public enum PinType {
+        TEXT,
+        IMAGE
+    }
+    private static final String TAG = Pin.class.getName();
+    private String authorUID;
+    private String caption;
+    private Date timestamp;
+    // contents will be a text if PinType == TEXT
+    // or a URL to an image if PinType == IMAGE
+    private String contents;
+    private PinType type;
+
+    public Pin(String caption, String authorUID, PinType type, String contents) {
+        this.caption = caption;
+        this.authorUID = authorUID;
+        this.type = type;
+        this.contents = contents;
+        this.timestamp = new Date();
+    }
+
+    // bruh I hate Java
+    public String getAuthorID() { return authorUID; }
+    public void setAuthorID(int authorID) { this.authorUID = authorUID; }
+    public String getCaption() { return caption; }
+    public void setCaption(String caption) { this.caption = caption; }
+    public String getContents() { return contents; }
+    public void setContents(String contents) { this.contents = contents; }
+    public PinType getType() { return type; }
+    public void setType(PinType type) { this.type = type; }
+
+    public void save() {
+        // handcrafting a map here instead of using reflection in Java
+        // since I was getting weird auth errors otherwise.
+        Map<String, Object> pin = new HashMap<>();
+        pin.put("caption", this.caption);
+        pin.put("authorUID", this.authorUID);
+        pin.put("contents", this.contents);
+        pin.put("type", String.valueOf(this.type));
+
+        Log.d(TAG, "Saving the following pin to firestore: " + pin.toString());
+
+        FirebaseFirestore.getInstance()
+            .collection("pins")
+            .add(pin)
+            // dat error handling tho
+            .addOnSuccessListener(new OnSuccessListener<DocumentReference>() {
+                @Override
+                public void onSuccess(DocumentReference documentReference) {
+                    Log.d(TAG, "DocumentSnapshot written with ID: " + documentReference.getId());
+                }
+            })
+            .addOnFailureListener(new OnFailureListener() {
+                @Override
+                public void onFailure(@NonNull Exception e) {
+                    Log.w(TAG, "Error adding document", e);
+                }
+            });
```

Comment:
If we end up removing this we probably don't need any of the logging messages either since the calling context will handle starting and ending the save transaction and how to surface the status.

Reply:
Change has been made. Error now propagates up to caller, which ATM will create a toast based on the return type.